aarch64/src/mailbox.rs

```diff
@@ -1,76 +1,287 @@
-pub const MBOX_STATUS: u64 = 0x18;
-pub const MBOX_WRITE: u64 = 0x20;
+const MBOX_READ: u64 = 0x00;
+const MBOX_STATUS: u64 = 0x18;
+const MBOX_WRITE: u64 = 0x20;
 
-pub const MBOX_FULL: u32 = 0x8000_0000;
-pub const MBOX_EMPTY: u32 = 0x4000_0000;
+const MBOX_FULL: u32 = 0x8000_0000;
+const MBOX_EMPTY: u32 = 0x4000_0000;
 
-#[repr(u32)]
-pub enum TagId {
-    GetClockRate = 0x38002,
+static MAILBOX: Lock<Option<&'static mut Mailbox>> = Lock::new("mailbox", None);
+
+/// Mailbox init.  Mainly initialises a lock to ensure only one mailbox request
+/// can be made at a time.  We have no heap at this point, so creating a mailbox
+/// that can be initialised based off the devicetree is rather convoluted.
+pub fn init(dt: &DeviceTree) {
+    static mut NODE: LockNode = LockNode::new();
+    let mut mailbox = MAILBOX.lock(unsafe { &NODE });
+    *mailbox = Some({
+        static mut MAYBE_MAILBOX: MaybeUninit<Mailbox> = MaybeUninit::uninit();
+        unsafe {
+            MAYBE_MAILBOX.write(Mailbox::new(dt));
+            MAYBE_MAILBOX.assume_init_mut()
+        }
+    });
+}
+
+/// https://developer.arm.com/documentation/ddi0306/b/CHDGHAIG
+/// https://github.com/raspberrypi/firmware/wiki/Mailbox-property-interface
+struct Mailbox {
+    reg: RegBlock,
+}
+
+impl Mailbox {
+    fn new(dt: &DeviceTree) -> Mailbox {
+        Mailbox {
+            reg: dt
+                .find_compatible("brcm,bcm2835-mbox")
+                .next()
+                .and_then(|uart| dt.property_translated_reg_iter(uart).next())
+                .and_then(|reg| reg.regblock())
+                .unwrap(),
+        }
+    }
+
+    fn request<T, U>(&self, req: &mut Message<T, U>)
+    where
+        T: Copy,
+        U: Copy,
+    {
+        // Read status register until full flag not set
+        while (read_reg(self.reg, MBOX_STATUS) & MBOX_FULL) != 0 {}
+
+        // Write the request address combined with the channel to the write register
+        let channel = ChannelId::ArmToVc as u32;
+        let uart_mbox_u32 = req as *const _ as u32;
+        let r = (uart_mbox_u32 & !0xF) | channel;
+        write_reg(self.reg, MBOX_WRITE, r);
+
+        // Wait for response
+        // FIXME: two infinite loops - can go awry
+        loop {
+            while (read_reg(self.reg, MBOX_STATUS) & MBOX_EMPTY) != 0 {}
+            let response = read_reg(self.reg, MBOX_READ);
+            if response == r {
+                break;
+            }
+        }
+    }
 }
 
 #[repr(u8)]
-pub enum ChannelId {
+enum ChannelId {
     ArmToVc = 8,
 }
 
 #[repr(C)]
 #[derive(Debug, Clone, Copy)]
-struct SetClockRateRequest {
+struct Request<T> {
     size: u32, // size in bytes
     code: u32, // request code (0)
+    tags: T,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Response<T> {
+    size: u32, // size in bytes
+    code: u32, // response code
+    pub tags: T,
+}
 
-    // Tag
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Tag<T> {
     tag_id0: u32,
```

Comment:
Can we type this so that only valid tags would be allowed? It should be possible.

Reply:
worth a look, but I'd leave for a followup.